mp2-v1/src/values_extraction/leaves/mapping_var.rs

```diff
@@ -0,0 +1,258 @@
+        deserialize_with = "deserialize_long_array"
+    )]
+    pub(crate) table_info: [ColumnInfo; MAX_COLUMNS],
+}
+
+impl<const NODE_LEN: usize, const MAX_COLUMNS: usize, const MAX_FIELD_PER_EVM: usize>
+    LeafMappingVarCircuit<NODE_LEN, MAX_COLUMNS, MAX_FIELD_PER_EVM>
+where
+    [(); PAD_LEN(NODE_LEN)]:,
+{
+    pub fn build(
+        b: &mut CBuilder,
+    ) -> LeafMappingVarWires<NODE_LEN, MAX_COLUMNS, MAX_FIELD_PER_EVM> {
+        let zero = b.zero();
+
+        let key_id = b.add_virtual_target();
+        let evm_word = b.add_virtual_target();
+        let table_info = array::from_fn(|_| b.add_virtual_column_info());
+        let [is_actual_columns, is_extracted_columns] =
+            array::from_fn(|_| array::from_fn(|_| b.add_virtual_bool_target_safe()));
+
+        let slot = MappingSlot::mpt_key_with_offset(b, evm_word);
+
+        // Range check for the slot to restrict it's an Uint8.
+        b.range_check(slot.mapping_slot, 8);
+        // Range check for the EVM word to restrict it's an Uint32.
+        b.range_check(evm_word, 32);
+
+        // Build the node wires.
+        let wires =
+            MPTLeafOrExtensionNode::build_and_advance_key::<_, D, NODE_LEN, MAX_LEAF_VALUE_LEN>(
+                b,
+                &slot.keccak_mpt.mpt_key,
+            );
+        let node = wires.node;
+        let root = wires.root;
+
+        // Left pad the leaf value.
+        let value: Array<Target, MAPPING_LEAF_VALUE_LEN> = left_pad_leaf_value(b, &wires.value);
+
+        // Compute the metadata digest.
+        let metadata_digest = MetadataGadget::<_, MAX_FIELD_PER_EVM>::new(
+            &table_info,
+            &is_actual_columns,
+            &is_extracted_columns,
+            evm_word,
+            slot.mapping_slot,
+        )
+        .build(b);
+
+        // key_column_md = H( "KEY" || slot)
+        let inputs = KEY_ID_PREFIX
+            .iter()
+            .map(|u| b.constant(F::from_canonical_u8(*u)))
+            .chain(once(slot.mapping_slot))
+            .collect();
+        let key_column_md = b.hash_n_to_hash_no_pad::<CHasher>(inputs);
+        // Add the information related to the key to the metadata.
+        // metadata_digest += D(key_column_md || key_id)
+        let inputs = key_column_md
+            .to_targets()
+            .into_iter()
+            .chain(once(key_id))
+            .collect_vec();
+        let metadata_key_digest = b.map_to_curve_point(&inputs);
+        let metadata_digest = b.add_curve_point(&[metadata_digest, metadata_key_digest]);
+
+        // Compute the values digest.
+        let values_digest = ColumnGadget::<MAX_FIELD_PER_EVM>::new(
+            &value.arr,
+            &table_info[..MAX_FIELD_PER_EVM],
+            &is_extracted_columns[..MAX_FIELD_PER_EVM],
+        )
+        .build(b);
+
+        // values_digest += D(key_id || left_pad32(key))
+        let inputs: Vec<_> = iter::once(key_id).chain(slot.mapping_key.arr).collect();
```

Comment:
Could we pack to u32 limbs before hashing it? This will allow to save field elements to be hashed, and will be easier in merkle-tree construction circuits to hash again cells values which are expected to be `U256`